File: core/src/main/kotlin/com/example/toyracers/surface/SurfaceSpeedConfig.kt
package com.example.toyracers.surface

/** Gradual speed-limit transition between road and off-road surfaces. */
data class SurfaceSpeedConfig(
    val offRoadSpeedMultiplier: Float = 0.3f,
    val transitionSeconds: Float = 3f,
) {
    init {
        require(offRoadSpeedMultiplier in 0f..1f)
        require(transitionSeconds > 0f)
    }
}

File: dart/lib/simulation/surface/surface_speed_config.dart
import '../math/float32.dart';

/// Gradual speed-limit transition between road and off-road surfaces.
final class SurfaceSpeedConfig {
  SurfaceSpeedConfig({
    double offRoadSpeedMultiplier = 0.3,
    double transitionSeconds = 3,
  }) : offRoadSpeedMultiplier = Float32.narrow(offRoadSpeedMultiplier),
       transitionSeconds = Float32.narrow(transitionSeconds) {
    if (this.offRoadSpeedMultiplier < 0 || this.offRoadSpeedMultiplier > 1) {
      throw ArgumentError.value(
        offRoadSpeedMultiplier,
        'offRoadSpeedMultiplier',
        'must be in the range [0, 1]',
      );
    }
    if (this.transitionSeconds <= 0) {
      throw ArgumentError.value(
        transitionSeconds,
        'transitionSeconds',
        'must be greater than zero',
      );
    }
  }

  final double offRoadSpeedMultiplier;
  final double transitionSeconds;
}
